sql/src/main/java/org/apache/druid/sql/calcite/schema/MetadataSegmentView.java

```diff
@@ -143,10 +146,16 @@ private void poll()
 
     final DateTime timestamp = DateTimes.nowUtc();
     while (metadataSegments.hasNext()) {
-      final DataSegment interned = DataSegmentInterner.intern(metadataSegments.next());
+      final SegmentWithOvershadowedStatus segment = metadataSegments.next();
+      final DataSegment interned = DataSegmentInterner.intern(segment.getDataSegment());
+      final SegmentWithOvershadowedStatus segmentWithOvershadowedStatus = new SegmentWithOvershadowedStatus(
+          interned,
+          segment.isOvershadowed()
+      );
       // timestamp is used to filter deleted segments
-      publishedSegments.put(interned, timestamp);
```

Comment:
This introduces a bug: since there are two possible SegmentWithOvershadowedStatus for each underlying DataSegment, now the same segment can be in `publishedSegments` twice for a period of time. There's a few ways to deal with this:

1. Make `publishedSegments` a `TreeSet<SegmentWithOvershadowedStatus>` and update the entire map atomically. This is a super clean solution but would burst to higher memory usage (it would need to keep two entire copies of the map in memory when replacing them).
2. Make `publishedSegments` a `ConcurrentSkipListMap<DataSegment, CachedSegmentInfo>` where CachedSegmentInfo is some static class, defined in this file, containing the updated timestamp and the overshadowed boolean. If you do this, the SegmentWithOvershadowedStatus won't be stored long term anymore. You could minimize memory footprint of CachedSegmentInfo, if you want, by making the timestamp a `long` rather than `DateTime`.
3. Make `publishedSegments` a `ConcurrentSkipListSet<SegmentWithOvershadowedStatus>`, make SegmentWithOvershadowedStatus mutable (in a thread-safe way), make its `equals`, `hashCode`, and `compareTo` methods only based on the `dataSegment` field, let its `overshadowed` field be modified, and add a `timestamp` field to it. When syncing the cache, get the current object and mutate the overshadowed field if necessary. Btw, a ConcurrentSkipListSet uses a ConcurrentSkipListMap under the hood, so the memory footprint savings of this aren't as much as you might expect relative to (2).

(2) is the variant that's closest to what the code was doing before this patch. One thing I don't love about it is that it is racey: it means that if a set of segments is overshadowed all at once, callers will not necessarily see a consistent view, because the map is being concurrently updated. They'll see the overshadowed flag get set for the underlying segments one at a time. But the same problem existed in the old code, so fixing it could be considered out of scope for this patch.

Reply:
If I was you I would probably do (2) and then consider if there's a way to address the raciness in a future patch.

Reply:
omg, thanks for catching this issue, actually I feel like going with option (1), i didn't do it initially because of memory bump it'd cause like you said, but that would have avoided such bugs and updates the cache atomically. I tried option (2) as well, one thing it would cause is, the return type of `getPublishedSegments()` changes, so `CachedSegmentInfo` cannot be private to this class. The `getPublishedSegments()` might need to be split into 2 methods `getPublishedSegments()` and `getCachedPublishedSegments()` or create yet another wrapper class and return that.  Another potential ugliness it might have is clients need to know if cache is enabled and call the right method to get published segments.

Reply:
okay, i added changes to do (2) now as we discussed.

Reply:
I don't have numbers, but I'm concerned about a `ConcurrentSkipListMap` of all-segments-in-system cardinality. Maybe take Gian's approach 1), but instead of `TreeMap` using sorted arrays of `DataSegment` objects. (Or Guava's `ImmutableSortedMap`, which uses the same approach underneath).

See https://webcache.googleusercontent.com/search?q=cache:csIsYj1a5oAJ:https://gist.github.com/gaul/7108880+&cd=2&hl=en&ct=clnk&gl=es&lr=lang_en%7Clang_ru:
```
ConcurrentSkipListMap    36
ImmutableSortedMap        8
```
So I'm pretty sure even temporarily having two `ImmutableSortedMap` in memory will well beat one `ConcurrentSkipListMap`.

Using sorted arrays directly, even that two maps can be avoided materializing in memory.

Reply:
I used `ImmutableSortedSet`, now the cache gets replaced atomically. @gianm @leventov let me know if you see any issues with this.

Reply:
Option 1 (what you went with) sounds good to me, especially in list of ConcurrentSkipListMap's additional overhead.